src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -516,6 +516,66 @@ class OrderedRVD(
 
   override def toUnpartitionedRVD: UnpartitionedRVD =
     new UnpartitionedRVD(typ.rowType, crdd)
+
+  def orderedLeftJoinDistinctAndInsert(
+    right: OrderedRVD,
+    root: String,
+    lift: Option[String] = None,
+    dropLeft: Option[Array[String]] = None
+  ): OrderedRVD = {
+    assert(!typ.key.contains(root))
+
+    val valueStruct = right.typ.valueType
+    val rightRowType = right.typ.rowType
+    val liftField = lift.map { name => rightRowType.field(name) }
+    val valueType = liftField.map(_.typ).getOrElse(valueStruct)
+
+    val removeLeft = dropLeft.map(_.toSet).getOrElse(Set.empty)
+    val keepIndices = rowType.fields.filter(f => !removeLeft.contains(f.name)).map(_.index).toArray
+    val newRowType = TStruct(
+      keepIndices.map(i => rowType.fieldNames(i) -> rowType.types(i)) ++ Array(root -> valueType): _*)
+
+    val localRowType = rowType
+
+    val shouldLift = lift.isDefined
+    val rightValueIndices = right.typ.valueIndices
+    val liftIndex = liftField.map(_.index).getOrElse(-1)
+
+    val joiner = { (ctx: RVDContext, it: Iterator[JoinedRegionValue]) =>
+      val rvb = ctx.rvb
+      val rv = RegionValue()
+
+      it.map { jrv =>
+        val lrv = jrv.rvLeft
+        val rrv = jrv.rvRight
+        rvb.start(newRowType)
+        rvb.startStruct()
+        rvb.addFields(localRowType, lrv, keepIndices)
+        if (rrv == null)
+          rvb.setMissing()
+        else {
+          if (shouldLift)
+            rvb.addField(rightRowType, rrv, liftIndex)
+          else {
+            rvb.startStruct()
+            rvb.addFields(rightRowType, rrv, rightValueIndices)
+            rvb.endStruct()
+          }
+        }
+        rvb.endStruct()
+        rv.set(ctx.region, rvb.end())
+        rv
+      }
+    }
+    keyBy(typ.key.take(right.typ.key.length)).orderedJoinDistinct(
+      right.keyBy(),
+      "left",
+      joiner,
+      typ.copy(rowType = newRowType,
+        partitionKey = typ.partitionKey.filter(!removeLeft.contains(_)),
+        key = typ.key.filter(!removeLeft.contains(_)))
```

Comment:
I feel like we should verify some invariants here and produce useful debugging error messages. In particular:

 - what if `typ.key` is shorter than `right.typ.key`? I think seeing the types printed will help me understand where the invariant breaking is occurring
 - it seems wrong to permit removing the key and partition keys?

Reply:
the way I'm using it, I do remove a suffix of the key (I remove the index field).

Reply:
I could also put this code inside the IR node.

Reply:
(that's the only place it's used)